File: tests/test_gradients.cpp
/**
 * Gradient verification via finite differences.
 * 
 * Compares AD-computed gradients against finite difference approximations
 * to verify correctness of the automatic differentiation implementation.
 */

#include <dmc/types.hpp>
#include <dmc/network.hpp>
#include <dmc/router.hpp>
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace dmc;

/**
 * Create a simple test network: 3 reaches in series.
 */
Network create_test_network() {
    Network net;
    
    // Three reaches in series
    for (int i = 0; i < 3; ++i) {
        Reach r;
        r.id = i;
        r.name = "reach_" + std::to_string(i);
        r.length = 5000.0;  // 5 km
        r.slope = 0.001;
        r.manning_n = Real(0.035);
        r.upstream_junction_id = (i > 0) ? i - 1 : -1;
        r.downstream_junction_id = i;
        net.add_reach(r);
        
        Junction j;
        j.id = i;
        j.upstream_reach_ids = {i};  // This reach flows INTO this junction
        if (i < 2) j.downstream_reach_ids = {i + 1};  // Next reach flows OUT
        j.is_headwater = (i == 0);
        j.is_outlet = (i == 2);
        net.add_junction(j);
    }
    
    net.build_topology();
    return net;
}

/**
 * Reset network state to initial conditions.
 */
void reset_network_state(Network& net) {
    for (int reach_id : net.topological_order()) {
        Reach& reach = net.get_reach(reach_id);
        reach.inflow_prev = Real(0.0);
        reach.inflow_curr = Real(0.0);
        reach.outflow_prev = Real(0.0);
        reach.outflow_curr = Real(0.0);
        reach.lateral_inflow = Real(0.0);
    }
}

/**
 * Run simulation and return outlet discharge.
 */
double run_simulation(Network& net, int num_steps, double dt) {
    // Reset state before each run
    reset_network_state(net);
    
    RouterConfig config;
    config.dt = dt;
    config.enable_gradients = false;
    
    MuskingumCungeRouter router(net, config);
    
    // Set constant lateral inflow
    for (int reach_id : net.topological_order()) {
        router.set_lateral_inflow(reach_id, 1.0);  // 1 m³/s per reach
    }
    
    // Run simulation
    for (int t = 0; t < num_steps; ++t) {
        router.route_timestep();
    }
    
    // Return outlet discharge (last reach)
    int outlet_id = net.topological_order().back();
    return router.get_discharge(outlet_id);
}

/**
 * Compute gradient via finite differences.
 */
double finite_difference_gradient(Network& net, int reach_id, 
                                   const std::string& param_name,
                                   int num_steps, double dt, double eps = 1e-5) {
    Reach& reach = net.get_reach(reach_id);
    
    // Get pointer to parameter
    Real* param = nullptr;
    if (param_name == "manning_n") {
        param = &reach.manning_n;
    } else if (param_name == "width_coef") {
        param = &reach.geometry.width_coef;
    } else {
        throw std::runtime_error("Unknown parameter: " + param_name);
    }
    
    // Save original value
    double orig = to_double(*param);
    
    // Forward perturbation
    *param = Real(orig + eps);
    double Q_plus = run_simulation(net, num_steps, dt);
    
    // Backward perturbation
    *param = Real(orig - eps);
    double Q_minus = run_simulation(net, num_steps, dt);
    
    // Restore original
    *param = Real(orig);
    
    // Central difference
    return (Q_plus - Q_minus) / (2.0 * eps);
}

/**
 * Compute gradient via automatic differentiation.
 */
double ad_gradient(Network& net, int reach_id, const std::string& param_name,
                   int num_steps, double dt) {
    if (!AD_ENABLED) {
        std::cerr << "AD not enabled, cannot compute AD gradient\n";
        return 0.0;
    }
    
    // Reset state before AD run (critical for matching FD!)
    reset_network_state(net);
    
    RouterConfig config;
    config.dt = dt;
    config.enable_gradients = true;
    
    MuskingumCungeRouter router(net, config);
    
    // Reset any previous AD tape state
    router.reset_gradients();
    
    // Set constant lateral inflow
    for (int rid : net.topological_order()) {
        router.set_lateral_inflow(rid, 1.0);
    }
    
    // Start recording
    router.start_recording();
    
    // Run simulation
    for (int t = 0; t < num_steps; ++t) {
        router.route_timestep();
    }
    
    // Stop recording
    router.stop_recording();
    
    // Compute gradients (dL/dQ = 1 for single output)
    int outlet_id = net.topological_order().back();
    std::vector<int> gauges = {outlet_id};
    std::vector<double> dL_dQ = {1.0};
    router.compute_gradients(gauges, dL_dQ);
    
    // Get gradient
    auto grads = router.get_gradients();
    std::string key = "reach_" + std::to_string(reach_id) + "_" + param_name;
    
    double grad = grads.count(key) ? grads.at(key) : 0.0;
    
    // Clean up
    router.reset_gradients();
    
    return grad;
}

int main() {
    std::cout << "=== Gradient Verification Test ===\n\n";
    
    if (!AD_ENABLED) {
        std::cout << "WARNING: AD is not enabled. Only finite difference test will run.\n";
    }
    
    Network net = create_test_network();
    
    int num_steps = 50;  // Increased to allow flow propagation through network
    double dt = 3600.0;  // 1 hour
    
    std::cout << std::fixed << std::setprecision(8);
    std::cout << "Configuration:\n";
    std::cout << "  Timesteps: " << num_steps << "\n";
    std::cout << "  dt: " << dt << " s\n";
    std::cout << "  Reaches: " << net.num_reaches() << "\n\n";
    
    // Test parameters
    std::vector<std::string> params = {"manning_n", "width_coef"};
    
    bool all_passed = true;
    double tol = 0.05;  // 5% relative tolerance (realistic for FD vs AD comparison)
    
    for (int reach_id : net.topological_order()) {
        for (const auto& param : params) {
            double fd_grad = finite_difference_gradient(net, reach_id, param, 
                                                         num_steps, dt);
            double ad_grad = 0.0;
            
            if (AD_ENABLED) {
                ad_grad = ad_gradient(net, reach_id, param, num_steps, dt);
            }
            
            double rel_error = 0.0;
            double abs_diff = std::abs(ad_grad - fd_grad);
            
            if (std::abs(fd_grad) > 1e-6) {
                rel_error = abs_diff / std::abs(fd_grad);
            } else if (std::abs(ad_grad) > 1e-6) {
                rel_error = 1.0;  // FD is zero but AD is not
            }
            // If both are very small, consider it a pass
            
            // Pass if: relative error < tol OR both gradients are negligibly small
            bool passed = (rel_error < tol) || (abs_diff < 1e-6) || !AD_ENABLED;
            
            std::cout << "Reach " << reach_id << ", " << param << ":\n";
            std::cout << "  FD gradient:  " << fd_grad << "\n";
            if (AD_ENABLED) {
                std::cout << "  AD gradient:  " << ad_grad << "\n";
                std::cout << "  Rel. error:   " << rel_error << "\n";
                std::cout << "  Status:       " << (passed ? "PASS" : "FAIL") << "\n";
            }
            std::cout << "\n";
            
            if (!passed) all_passed = false;
        }
    }
    
    if (AD_ENABLED) {
        std::cout << "=== Overall: " << (all_passed ? "ALL TESTS PASSED" : "SOME TESTS FAILED") 
                  << " ===\n";
        return all_passed ? 0 : 1;
    } else {
        std::cout << "=== FD-only test complete (AD not enabled) ===\n";
        return 0;
    }
}

File: tests/python/test_gradients.py
"""Test gradient computation functionality."""

import pytest
import numpy as np


def create_simple_network():
    """Create a simple 3-reach linear network for testing."""
    import droute

    network = droute.Network()

    # Add junctions first (required for MuskingumCungeRouter)
    for i in range(4):
        junction = droute.Junction()
        junction.id = i
        network.add_junction(junction)

    for i in range(3):
        reach = droute.Reach()
        reach.id = i
        reach.length = 5000.0
        reach.slope = 0.001
        reach.manning_n = 0.035
        reach.upstream_junction_id = i
        reach.downstream_junction_id = i + 1
        network.add_reach(reach)

    network.build_topology()
    return network


class TestMuskingumCungeGradients:
    """Test gradient computation for MuskingumCungeRouter."""

    def test_gradient_recording(self):
        """Test starting and stopping gradient recording."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()

        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
            router.record_output(2)

        router.stop_recording()

        # Should have recorded 5 outputs
        assert router.get_output_history_size(2) == 5

    def test_compute_gradients(self):
        """Test computing gradients."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()

        n_timesteps = 10
        for _ in range(n_timesteps):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
            router.record_output(2)

        router.stop_recording()

        # Compute gradients with unit gradient
        dL_dQ = [1.0] * n_timesteps
        router.compute_gradients_timeseries(2, dL_dQ)

        grads = router.get_gradients()
        assert isinstance(grads, dict)

    def test_reset_gradients(self):
        """Test resetting gradients."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
            router.record_output(2)
        router.stop_recording()

        router.compute_gradients_timeseries(2, [1.0] * 5)
        router.reset_gradients()

        # After reset, get_gradients should still work
        grads = router.get_gradients()
        assert isinstance(grads, dict)

    def test_clear_output_history(self):
        """Test clearing output history."""
        import droute

        network = create_simple_network()
        router = droute.MuskingumCungeRouter(network, droute.RouterConfig())

        router.start_recording()
        for _ in range(5):
            router.route_timestep()
            router.record_output(2)
        router.stop_recording()

        assert router.get_output_history_size(2) == 5

        router.clear_output_history()
        assert router.get_output_history_size(2) == 0

    def test_get_output_history(self):
        """Test retrieving output history."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()
        for t in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
            router.record_output(2)
        router.stop_recording()

        history = router.get_output_history(2)
        assert len(history) == 5
        # All values should be non-negative
        assert all(q >= 0 for q in history)

    @pytest.mark.slow
    def test_gradient_finite_difference_check(self):
        """Verify AD gradients match finite differences."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        config.dt = 3600.0

        outlet_id = 2
        n_timesteps = 20
        eps = 1e-5

        # Helper to run simulation and get loss
        def run_simulation(manning_n_vals):
            net = create_simple_network()
            # Set Manning's n values
            for i, n in enumerate(manning_n_vals):
                net.get_reach(i).manning_n = n

            router = droute.MuskingumCungeRouter(net, config)

            total_loss = 0.0
            for t in range(n_timesteps):
                for i in range(3):
                    router.set_lateral_inflow(i, 10.0 + t * 0.5)
                router.route_timestep()
                q = router.get_discharge(outlet_id)
                # Simple loss: sum of squared outputs
                total_loss += q * q

            return total_loss

        # Base Manning's n
        base_n = [0.035, 0.035, 0.035]
        base_loss = run_simulation(base_n)

        # Compute numerical gradient for first reach
        perturbed_n = base_n.copy()
        perturbed_n[0] = base_n[0] + eps
        loss_plus = run_simulation(perturbed_n)

        perturbed_n[0] = base_n[0] - eps
        loss_minus = run_simulation(perturbed_n)

        numerical_grad = (loss_plus - loss_minus) / (2 * eps)

        # The gradient should be non-zero and finite
        assert np.isfinite(numerical_grad)
        # For this simple case, changing manning_n should affect loss
        assert abs(numerical_grad) > 0


class TestIRFGradients:
    """Test gradient computation for IRFRouter."""

    def test_irf_gradient_recording(self):
        """Test IRF router gradient recording."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.IRFRouter(network, config)

        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
        router.stop_recording()

        # Should complete without error
        router.compute_gradients([2], [1.0])
        grads = router.get_gradients()
        assert isinstance(grads, dict)


class TestDiffusiveWaveGradients:
    """Test gradient computation for DiffusiveWaveRouter."""

    def test_dw_gradient_recording(self):
        """Test DiffusiveWave router gradient recording."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.DiffusiveWaveRouter(network, config)

        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
        router.stop_recording()

        router.compute_gradients([2], [1.0])
        grads = router.get_gradients()
        assert isinstance(grads, dict)


class TestDiffusiveWaveIFTGradients:
    """Test IFT-based gradients for DiffusiveWaveIFT."""

    def test_ift_gradient_computation(self):
        """Test IFT adjoint gradient computation."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        config.dw_use_ift_adjoint = True
        router = droute.DiffusiveWaveIFT(network, config)

        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
        router.stop_recording()

        router.compute_gradients([2], [1.0])
        grads = router.get_gradients()
        assert isinstance(grads, dict)


class TestSoftGatedKWTGradients:
    """Test gradient computation for SoftGatedKWT."""

    def test_soft_kwt_gradients(self):
        """Test SoftGatedKWT gradient recording and computation."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.SoftGatedKWT(network, config)

        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
        router.stop_recording()

        router.compute_gradients([2], [1.0])
        grads = router.get_gradients()
        assert isinstance(grads, dict)


class TestNetworkGradients:
    """Test network-level gradient operations."""

    def test_network_zero_gradients(self):
        """Test zeroing network gradients."""
        import droute

        network = create_simple_network()
        network.zero_gradients()  # Should not raise

    def test_get_grad_manning_n_all(self):
        """Test getting all Manning's n gradients."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        # Run and compute gradients
        router.start_recording()
        for _ in range(5):
            for i in range(3):
                router.set_lateral_inflow(i, 10.0)
            router.route_timestep()
            router.record_output(2)
        router.stop_recording()

        router.compute_gradients_timeseries(2, [1.0] * 5)

        # Get gradients from network
        grads = network.get_grad_manning_n_all()
        assert len(grads) == 3


class TestGradientEdgeCases:
    """Test edge cases in gradient computation."""

    def test_gradients_with_zero_inflow(self):
        """Test gradients when inflow is zero."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()
        for _ in range(5):
            # Zero inflow
            for i in range(3):
                router.set_lateral_inflow(i, 0.0)
            router.route_timestep()
            router.record_output(2)
        router.stop_recording()

        # Should still work (even if gradients are zero)
        router.compute_gradients_timeseries(2, [1.0] * 5)
        grads = router.get_gradients()
        assert isinstance(grads, dict)

    def test_gradients_with_single_timestep(self):
        """Test gradients with just one timestep."""
        import droute

        network = create_simple_network()
        config = droute.RouterConfig()
        config.enable_gradients = True
        router = droute.MuskingumCungeRouter(network, config)

        router.start_recording()
        for i in range(3):
            router.set_lateral_inflow(i, 10.0)
        router.route_timestep()
        router.record_output(2)
        router.stop_recording()

        router.compute_gradients_timeseries(2, [1.0])
        grads = router.get_gradients()
        assert isinstance(grads, dict)
